src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs

```diff
@@ -11,65 +12,107 @@ internal static class RestoreSqlBackupImpl
+                mainCommand.ExecuteNonQuery();
+                if (hasDifferentialBackup)
+                {
+                    var differentialCommand = GetSqlCommand(context, connection, settings, differentialBackupFiles, settings.DifferentialBackupSetFile, false, false);
+                    differentialCommand.ExecuteNonQuery();
+                }
+            }
+        }
 
-                var logicalNames = GetLogicalNames(backupFile, connection);
+        private static SqlCommand GetSqlCommand(ICakeContext context, SqlConnection connection, RestoreSqlBackupSettings settings, IList<FilePath> backupFiles, int backupSetFile = 1, bool noRecovery = false, bool allowAlterDb = true)
+        {
+            var firstBackupFile = backupFiles.First();
+            var oldDbName = GetDatabaseName(firstBackupFile, connection);
+            var newDatabaseName = settings.NewDatabaseName ?? oldDbName;
+            context.Log.Information($"Using database name '{newDatabaseName}' to be a name for the restored database");
 
-                var sql = "";
+            var logicalNames = GetLogicalNames(firstBackupFile, connection);
 
-                if (settings.SwitchToSingleUserMode)
-                {
-                    sql += $@"
+            var sql = "";
+
+            if (settings.SwitchToSingleUserMode && allowAlterDb)
+            {
+                sql += $@"
 if db_id({Sql.EscapeNameQuotes(newDatabaseName)}) is not null 
 begin
     use master;
     alter database {Sql.EscapeName(newDatabaseName)} set single_user with rollback immediate;
 end
 ";
+            }
+
+            sql += $"Restore database {Sql.EscapeName(newDatabaseName)} from\r\n";
+            for (var i = 0; i < backupFiles.Count; i++)
+            {
+                sql += $" disk = @backupFile{i}";
+                if (i < backupFiles.Count - 1)
+                {
+                    sql += ","; // only need comma before penultimate list
                 }
 
-                sql += $"Restore database {Sql.EscapeName(newDatabaseName)} from disk = @backupFile with ";
+                sql += "\r\n";
+            }
 
-                for (var i = 0; i < logicalNames.Count; i++)
+            sql += $" with file = {backupSetFile},\r\n";
+            for (var i = 0; i < logicalNames.Count; i++)
+            {
+                sql += $" move @LName{i} to @LPath{i}";
+                if (i < logicalNames.Count - 1)
                 {
-                    sql += $" move @LName{i} to @LPath{i}";
-                    if (i < logicalNames.Count - 1)
-                    {
-                        sql += ", \r\n"; // only need comma before penultimate list
-                    }
+                    sql += ",\r\n"; // only need comma before penultimate list
                 }
-                if (settings.WithReplace)
+            }
+
+            if (settings.WithReplace)
```

Comment:
Replace is not really applicable to diff backups, yet it will crop in if specified in options

Reply:
The specifics of the parameters of the individual database restore commands are now no longer handled inside `GetRestoreSqlBackupCommand`. This also solves the issue mentioned here.